librz/bin/pdb/tpi.c

```diff
@@ -1,778 +1,356 @@
 // SPDX-FileCopyrightText: 2014 inisider <inisider@gmail.com>
+// SPDX-FileCopyrightText: 2021 Basstorm <basstorm@nyist.edu.cn>
 // SPDX-License-Identifier: LGPL-3.0-only
 
-#include "types.h"
-#include "tpi.h"
-#include "stream_file.h"
+#include "pdb.h"
 
-// TODO: remove these global variables
-static unsigned int base_idx = 0;
-static RzList *p_types_list;
-
-static bool is_simple_type(int idx) {
-	ut32 value = (ut32)idx;
+static bool is_simple_type(TpiStream *stream, ut32 idx) {
 	/*   https://llvm.org/docs/PDB/TpiStream.html#type-indices
-        .---------------------------.------.----------.
-        |           Unused          | Mode |   Kind   |
-        '---------------------------'------'----------'
-        |+32                        |+12   |+8        |+0
-	*/
-	return value < base_idx;
-	// return ((value & 0x00000000FFF00) <= 0x700 && (value & 0x00000000000FF) < 0x80);
+  .---------------------------.------.----------.
+  |           Unused          | Mode |   Kind   |
+  '---------------------------'------'----------'
+  |+32                        |+12   |+8        |+0
+  */
+	return idx < stream->header.TypeIndexBegin;
+	// return ((value & 0x00000000FFF00) <= 0x700 && (value & 0x00000000000FF) <
+	// 0x80);
 }
 
-static int skip_padding(uint8_t *leaf_data, unsigned int *read_bytes, unsigned int len) {
-	unsigned int c = 0;
-	CAN_READ(*read_bytes, 1, len);
-	while ((*leaf_data & 0xf0) == 0xf0) {
-		c++;
-		CAN_READ((*read_bytes + c), 1, len);
-		leaf_data++;
-	}
-	(*read_bytes) += (c);
-	return c;
+int tpi_type_node_cmp(const void *incoming, const RBNode *in_tree, void *user) {
+	ut64 ia = *(ut64 *)incoming;
```

Comment:
@thestr4ng3r @Basstorm shouldn't this be 
```suggestion
int tpi_type_node_cmp(const void *incoming, const RBNode *in_tree, void *user) {
	ut64 ia = ((const TpiType *)incoming)->type_index;
```
?

Reply:
No. We will get nothing from `RBTree` if in this way.

Reply:
`incoming` here, at least from the insertion in `parse_tpi_stream` is a `ut32` though, as it is just a pointer to `TpiType.type_index` so it should probably be `ut32 ia = *(ut32 *)incoming;`

Reply:
I see in another place it is used with `ut64`. This should be unified to a single type I guess.

Reply:
The `RBTree` issue is solved by `ut32 ia = *(ut32 *)incoming;` :)